### PR TITLE
Support gRPC richer error model

### PR DESCRIPTION
**Pull request by jev-e**
# Description

- Provide a mechanism for users of the .NET SDK to obtain a Dapr implementation of the richer error model and extended error information. 

- Add an extension to DaprExtension that attempts to retrieve a `DaprExtendedErrorInfo`, which contains a collection of 
`DaprExtendedErrorDetail` each of which is a mirror of its gRPC specification, and provides all properties provided by said specification. 

- Add associated unit tests.

- Update daprdocs in relation to the .NET SDK, providing an error handling section, which in turn provides an explanation of the richer error model, links to further reading, and details implementation in .NET SDK and example usage.

## Issue reference

#1228 and #1227 

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [x] Created/updated tests
* [x] Extended the documentation

### CODE REVIEW

**Comment by WhitWaldo** [comment]
@jev-e Could you please make sure you've signed all your commits? The DCO check has failed.

**Comment by jev-e** [comment]
No clue why after the rebase its reporting so many new + changed files, it was ten beforehand.

**Comment by jev-e** [comment]
commit history and file weirdness should be fixed, just let github handle the master branch update.

**Comment by jev-e** [comment]
On the failing integration tests; I ran them locally and saw a similar failure. however after running the ones targeting /net 7 + then running the .net 6 again, they all pass successfully and I am unable to regenerate the same issue. Are there any known issues with these tests? I cant think what I have done (besides install commonProtos into dapr.common) that would cause a failing GRPC integration test.

File: src/Dapr.Common/Exceptions/ExtendedErrorDetailFactory.cs
Comment (jev-e) [inline src/Dapr.Common/Exceptions/ExtendedErrorDetailFactory.cs:17]: This is probably the only bit im actually iffy on; probably just reads better if its a switch assignment in the create

File: src/Dapr.Common/Exceptions/DaprExtendedErrorDetail.cs
Comment (jev-e) [inline src/Dapr.Common/Exceptions/DaprExtendedErrorDetail.cs:25]: the original proposal just returned null; given the dapr runtime does seem to return a UnknownDetailType if there is a [cast failure](https://github.com/dapr/kit/blob/main/errors/errors.go#L345); it makes sense to me to try and capture this and provide as much detail to the user as possible.

**Comment by WhitWaldo** [comment]
Don't worry about the failing integration tests - I'll call them out if they're an issue, but the test runner on GitHub routinely fails to connect to the endpoint in a timely manner and marks the whole attempt as a failure. They all have to pass to merge, but if you pass at least one of them, odds are great you'll pass them all before the re-run before merge.

**Comment by WhitWaldo** [comment]
@jev-e  I've got a few additional minor notes above if you don't mind taking a pass by them.

**Comment by WhitWaldo** [comment]
> On the failing integration tests; I ran them locally and saw a similar failure. however after running the ones targeting /net 7 + then running the .net 6 again, they all pass successfully and I am unable to regenerate the same issue. Are there any known issues with these tests? I cant think what I have done (besides install commonProtos into dapr.common) that would cause a failing GRPC integration test.

Don't worry about the failures. The test runner in GitHub often fails at least one of the integration tests (error logs indicate it just times out during setup) and it's easily re-run (as I just did). Usually if you pass at least one of the integration tests, you're good to go.

**Review by WhitWaldo — CHANGES_REQUESTED** [review]
Neglected to submit my review - apologies!

File: src/Dapr.Common/Exceptions/ExtendedErrorDetailFactory.cs
Comment (WhitWaldo) [reply src/Dapr.Common/Exceptions/ExtendedErrorDetailFactory.cs:17]: As it's an internal class and not publicly out there, I think it's sufficient to swap to a switch statement + a comment indicating it's intended to be a lookup table. I don't envision this code is going to need a lot of changes once merged given the maturity of gRPC.

File: src/Dapr.Common/Exceptions/DaprExtendedErrorDetail.cs
Comment (jev-e) [reply src/Dapr.Common/Exceptions/DaprExtendedErrorDetail.cs]: Changed IReadOnlyCollection

File: src/Dapr.Common/Exceptions/DaprExtendedErrorInfo.cs
Comment (jev-e) [reply src/Dapr.Common/Exceptions/DaprExtendedErrorInfo.cs:26]: changed to IReadOnlyCollection

File: src/Dapr.Common/Exceptions/DaprExtendedErrorDetail.cs
Comment (jev-e) [reply src/Dapr.Common/Exceptions/DaprExtendedErrorDetail.cs]: changed to IReadOnlyCollection

File: src/Dapr.Common/Exceptions/DaprExtendedErrorConstants.cs
Comment (jev-e) [reply src/Dapr.Common/Exceptions/DaprExtendedErrorConstants.cs:1]: Added copyright to all new (non-test) files.

File: daprdocs/content/en/dotnet-sdk-docs/dotnet-error-handling/dotnet-richer-error-model.md
Comment (jev-e) [reply daprdocs/content/en/dotnet-sdk-docs/dotnet-error-handling/dotnet-richer-error-model.md]: fixed

**Comment by WhitWaldo** [comment]
@jev-e This looks good to me - let me know if you want me to wait on merging it until you add integration tests/speak with the runtime maintainers or you want to tackle that in a separate PR later (perhaps as part of the 1.16 push).

**Review by WhitWaldo — APPROVED** [review]

File: daprdocs/content/en/dotnet-sdk-docs/dotnet-error-handling/_index.md
Comment (WhitWaldo) [inline daprdocs/content/en/dotnet-sdk-docs/dotnet-error-handling/_index.md:5]: Could you please change this weight to 90000? It should ideally show up under the various Dapr clients, but before Dev Integrations in the docs.

**Review by WhitWaldo — CHANGES_REQUESTED** [review]

File: daprdocs/content/en/dotnet-sdk-docs/dotnet-error-handling/_index.md
Comment (jev-e) [reply daprdocs/content/en/dotnet-sdk-docs/dotnet-error-handling/_index.md]: changed now

**Review by WhitWaldo — APPROVED** [review]

**Comment by WhitWaldo** [comment]
@jev-e Thank you very much for your contribution!

**Comment by marcduiker** [comment]
@holopin-bot @jev-e Thank you! Here's a digital badge as a small token of appreciation.

**Pull request merged**
